File: desktop/backend/src/diveintocrypto_desktop/engine/consensus/regime.py
"""Regime-Adaptive Weighting — strategy overlay.

Detects the market regime (TREND / RANGE / MIXED) from the already-computed ADX and
Choppiness outputs, then re-weights the indicator vote to emphasise the families that work
in that regime: trend/breakout indicators when trending, oscillators/mean-reversion when
ranging. This is ADDITIVE — it surfaces a regime label and an adaptively-weighted score
next to the parity-locked consensus; it never replaces the canonical verdict.
"""

from __future__ import annotations

from typing import Any

from diveintocrypto_desktop.engine.consensus.scorer import compute_weighted_score

# Indicator families by the regime they thrive in.
_TREND_FAMILY = {
    "macd", "ema_cross", "sma_cross", "ichimoku", "psar", "adx_di", "supertrend", "vortex",
    "aroon_oscillator", "schaff_trend_cycle", "trix", "kst", "coppock_curve", "linreg_slope",
    "kalman_trend", "donchian_breakout", "keltner_breakout", "vwma_cross", "hurst", "rolling_sharpe",
    "elder_ray", "wavetrend",
}
_RANGE_FAMILY = {
    "rsi", "stochastic", "williams_r", "cci", "bollinger", "bollinger_percent_b", "mfi",
    "connors_rsi", "stoch_rsi", "ultimate_oscillator", "zscore_reversion", "half_life_reversion",
    "relative_vigor_index", "fisher_transform", "cmo", "hist_vol_percentile",
}


def _raw(results: list, name: str, key: str) -> float | None:
    for r in results:
        if r.name == name and r.raw_values:
            v = r.raw_values.get(key)
            if isinstance(v, (int, float)):
                return float(v)
    return None


def detect_regime(results: list, cfg: dict | None = None) -> dict[str, Any]:
    cfg = (cfg or {}).get("regime", {})
    strong_adx = float(cfg.get("strong_adx", 25.0))
    weak_adx = float(cfg.get("weak_adx", 20.0))
    chop_trend = float(cfg.get("chop_trend", 38.2))
    chop_range = float(cfg.get("chop_range", 61.8))

    adx = _raw(results, "adx_di", "adx")
    chop = _raw(results, "choppiness", "chop")
    t = r = 0
    if adx is not None:
        if adx >= strong_adx:
            t += 1
        elif adx <= weak_adx:
            r += 1
    if chop is not None:
        if chop <= chop_trend:
            t += 1
        elif chop >= chop_range:
            r += 1
    regime = "TREND" if t > r else "RANGE" if r > t else "MIXED"
    return {"regime": regime, "adx": adx, "chop": chop}


def adaptive_weights(base_weights: dict, regime: str, boost: float = 1.5, damp: float = 0.6) -> dict:
    if regime == "TREND":
        up, dn = _TREND_FAMILY, _RANGE_FAMILY
    elif regime == "RANGE":
        up, dn = _RANGE_FAMILY, _TREND_FAMILY
    else:
        return dict(base_weights)
    out = dict(base_weights)
    for k, w in list(out.items()):
        if w == 0:  # keep pure filters (e.g. atr_filter) at 0
            continue
        if k in up:
            out[k] = w * boost
        elif k in dn:
            out[k] = w * damp
    return out


def evaluate(results: list, base_weights: dict, cfg: dict | None = None) -> dict[str, Any]:
    """Return regime + adaptively-weighted score (observational; consensus is untouched)."""
    reg = detect_regime(results, cfg)
    w = adaptive_weights(base_weights, reg["regime"])
    score = compute_weighted_score(results, w)
    return {
        "regime": reg["regime"],
        "adx": reg["adx"],
        "chop": reg["chop"],
        "adaptive_score": score["weighted_score"],
    }

File: android/app/src/commonMain/kotlin/com/diveintocrypto/android/domain/consensus/Regime.kt
package com.diveintocrypto.android.domain.consensus

import com.diveintocrypto.android.domain.model.IndicatorResult

/**
 * Regime-Adaptive Weighting — strategy overlay. Faithful port of the Python
 * reference `engine/consensus/regime.py`.
 *
 * Detects the market regime (TREND / RANGE / MIXED) from the already-computed ADX
 * and Choppiness outputs, then re-weights the indicator vote to emphasise the
 * families that work in that regime. ADDITIVE — it surfaces a regime label and an
 * adaptively-weighted score next to the parity-locked consensus; it never replaces
 * the canonical verdict.
 */
object Regime {

    data class Detection(val regime: String, val adx: Double?, val chop: Double?)

    data class Evaluation(
        val regime: String,
        val adx: Double?,
        val chop: Double?,
        val adaptiveScore: Double,
    )

    private val TREND_FAMILY = setOf(
        "macd", "ema_cross", "sma_cross", "ichimoku", "psar", "adx_di", "supertrend", "vortex",
        "aroon_oscillator", "schaff_trend_cycle", "trix", "kst", "coppock_curve", "linreg_slope",
        "kalman_trend", "donchian_breakout", "keltner_breakout", "vwma_cross", "hurst",
        "rolling_sharpe", "elder_ray", "wavetrend",
    )
    private val RANGE_FAMILY = setOf(
        "rsi", "stochastic", "williams_r", "cci", "bollinger", "bollinger_percent_b", "mfi",
        "connors_rsi", "stoch_rsi", "ultimate_oscillator", "zscore_reversion",
        "half_life_reversion", "relative_vigor_index", "fisher_transform", "cmo",
        "hist_vol_percentile",
    )

    private fun raw(results: List<IndicatorResult>, name: String, key: String): Double? =
        results.firstOrNull { it.name == name }?.rawValues?.get(key)

    fun detect(
        results: List<IndicatorResult>,
        strongAdx: Double = 25.0,
        weakAdx: Double = 20.0,
        chopTrend: Double = 38.2,
        chopRange: Double = 61.8,
    ): Detection {
        val adx = raw(results, "adx_di", "adx")
        val chop = raw(results, "choppiness", "chop")
        var t = 0; var r = 0
        if (adx != null) {
            if (adx >= strongAdx) t++ else if (adx <= weakAdx) r++
        }
        if (chop != null) {
            if (chop <= chopTrend) t++ else if (chop >= chopRange) r++
        }
        val regime = if (t > r) "TREND" else if (r > t) "RANGE" else "MIXED"
        return Detection(regime, adx, chop)
    }

    fun adaptiveWeights(
        baseWeights: Map<String, Double>,
        regime: String,
        boost: Double = 1.5,
        damp: Double = 0.6,
    ): Map<String, Double> {
        val (up, dn) = when (regime) {
            "TREND" -> TREND_FAMILY to RANGE_FAMILY
            "RANGE" -> RANGE_FAMILY to TREND_FAMILY
            else -> return baseWeights.toMap()
        }
        val out = HashMap<String, Double>(baseWeights.size)
        for ((k, w) in baseWeights) {
            out[k] = when {
                w == 0.0 -> w            // keep pure filters (e.g. atr_filter) at 0
                k in up -> w * boost
                k in dn -> w * damp
                else -> w
            }
        }
        return out
    }

    /** Regime + adaptively-weighted score (observational; consensus untouched). */
    fun evaluate(results: List<IndicatorResult>, baseWeights: Map<String, Double>): Evaluation {
        val det = detect(results)
        val w = adaptiveWeights(baseWeights, det.regime)
        val score = Scorer.compute(results, w)
        return Evaluation(det.regime, det.adx, det.chop, score.weightedScore)
    }
}
